keymap.c

```diff
@@ -1298,6 +1298,166 @@ enum CommandResult mutt_parse_bind(struct Buffer *buf, struct Buffer *s,
   return rc;
 }
 
+/**
+ * parse_menu - Parse a menu-names into array
+ * @param menu     Array for results
+ * @param s        string containing menu-names string
+ * @param maxmenus Total number of menus
+ * @param nummenus Number of menus this config applies to
+ * @param err      Buffer for error messages
+ * @param bind     If true 'bind', otherwise 'macro'
+ *
+ * Expects to see: <menu-string>[,<menu-string>]
+ */
+static void *parse_menu(int *menu, char *s, int maxmenus, int *nummenus,
+                        struct Buffer *err, char *cmd)
+{
+  int i = 0;
+  char *q = NULL;
+
+  /* menu name */
+  while (i < maxmenus)
+  {
+    q = strchr(s, ',');
+    if (q)
+      *q = '\0';
+
+    menu[i] = mutt_map_get_value(s, Menus);
+    if (menu[i] == -1)
+      mutt_buffer_printf(err, _("%s: no such menu"), s);
+    i++;
+    if (q)
+      s = q + 1;
+  }
+  *nummenus = i;
+  return NULL;
+}
```

Comment:
Stealing from #1543 (sorry, we have to use this old way until that's merged), we can remove the manual string splitting and `strsep()` instead.

```C
int i = 0;

char *menu_names_dup = mutt_str_strdup(s);
char *menu_name = strsep(&menu_names_dup, ",");

while (menu_name)
{
   menu[i] = mutt_map_get_value(menu_name, Menus);
   if (menu[i] == -1)
      mutt_buffer_printf(err, _("%s: no such menu"), menu_name);

   i++;
   menu_name = strsep(&menu_names_dup, ","); // This is keep assignment out of the conditional.
}

FREE(&menu_names_dup);

*nummenus = i;
return NULL;
```

This makes it a bit easier to understand what's happening (in my opinion of course, if you disagree please voice that.) However, this keeps two parts of the code I do not like:

First,
```C
menu[i] = mutt_map_get_value(menu_name, Menus);
if (menu[i] == -1)
   mutt_buffer_printf(err, _("%s: no such menu"), menu_name);
```
With this program flow, it's possible for `menu` to be `[-1, -1, -1, -1]` if there are four non-existent menu names. A static array does not seem like a good data structure choice here. We need something that grows so we can have code such as:

```C
int value = mutt_map_get_value(menu_name, Menus);

if (value != -1)
{
   // Add to data structure.
}
```

Second, what happens if we specify more menus than room in the `menu`?

We will run out of space in crash since we exceeded our allocated memory. 

The current way it's code has the opposite problem, if we specify more menus than `maxmenus`, it just stops parsing them. Try it: `unbind test,test,test,test,test,test,test,test,test,test,test,test,test,test,test,test,test,test,test,index q` doesn't work, but `unbind index q` does! A fixed array is not the best choice of a data structure in this case. We should use a list. I haven't had to use Neomutt's list implementation so this will be a learning experience for both of us :D 

@flatcap are you available to tutor us?

After we swap to a dynamic data structure, `i`, `nummenus`, and `maxmenus` are unnecessary since we can query the data structure for its size.

Reply:
>  if there are four non-existent menu names.

We seem to have lost a `return` statement after the error message.

> if we specify more menus than `maxmenus`, it just stops parsing them

An alternative way to do the parsing might be:
- create array of size `MENU_MAX`
- initialise to zero
- `x = mutt_map_get_value(...)`
- `array[x] = true;`

So rather than an array:
- `3, 5, 2, -1`
we have:
- `0, 1, 1, 0, 1, 0, 0, 0, 0`

Reply:
> An alternative way to do the parsing might be:

That's a clever approach and it should simplify things overall. It handles arbitrarily large inputs with a fixed amount of space.

So in `parse_menu()` becomes:

```C
char *menu_names_dup = mutt_str_strdup(s);
char *menu_name = strsep(&menu_names_dup, ",");

while (menu_name)
{
   int value = mutt_map_get_value(menu_name, Menus);
   if (value == -1)
   {
      mutt_buffer_printf(err, _("%s: no such menu"), menu_name);
      FREE(&menu_names_dup);
      return NULL; // this is just a placeholder.
   }
   else
   {
       menu[value] = true;
   }

   menu_name = strsep(&menu_names_dup, ","); 
}

FREE(&menu_names_dup);

return NULL;
```

If we drop the early return, we can handle cases where a correct menu follows an incorrect one e.g. `unbind test,pager q`. We would want to remove the `FREE()` as well.

Additionally, your approach simplifies `mutt_parse_unbind` as well. Instead of tracking a boolean for `all_menus`, we can set `menus` can be assigned to an array of `MAX_MENU` size with all values of `1`. It's `for` loop would become this:

```C
  for (int i = 0; i < MENU_MAX; ++i)
  {
    if (menu[i] != 1)
      continue; // Look what you made me type, flatcap. The horror you've wrought upon me

    int map = menu[i]; // This can be inlined or left as variable.
    if (all_keys)
    {
      km_unbind_all(&Keymaps[map],data);
      if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
      {
        km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
        km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
        km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
        km_bindkey("q", MENU_GENERIC, OP_EXIT);
        km_bindkey("q", MENU_PAGER, OP_EXIT);
        km_bindkey("?", MENU_GENERIC, OP_HELP);
      }
    }
    else
    {
      km_bindkey(key, map, OP_NULL);
    }
  }
```
which has a more uniform control flow.

Reply:
I kept return and free in parse_menu() until we agree what's the best. The same situation as with list of km_bindkeys in mutt_parse_unbind().